src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -452,9 +468,15 @@ entails SolverOptions{..} constraint context hints =
             -- @Coercible (T[X/a]) b = Coercible X b@ in the example).
             let wrappedTySub = replaceAllTypeVars (zip tvs xs) wrappedTy
             pure [Constraint nullSourceAnn C.Coercible [] [wrappedTySub, b] Nothing]
+        | RCons{} <- a
+        , RCons{} <- b -> do
+          let (constraints, (rl1, rl2)) = alignRowsWith (\a' b' -> Constraint nullSourceAnn C.Coercible [] [a', b'] Nothing) a b
+          unless (null (fst rl1) && null (fst rl2)) $
+            throwError . errorMessage $ TypesDoNotUnify (rowFromList rl1) (rowFromList rl2)
```

Comment:
I think we should add tests for coercing polymorphic rows.
```purescript
test :: forall r. { a :: Int | r } -> { a :: Id1 Int | r }
test = coerce
```
I don't think there's a reason this shouldn't coerce since the tails unify.

```purescript
test :: forall r s. { a :: Int | r } -> { a :: Id1 Int | s }
test = coerce
```
I'd think this needs to introduce a subgoal for `Coercible r s`, but `Coercible` isn't polykinded, so I assume we'll need to reject this with a unification error.

Reply:
Should we consider having special casing for records so that it introduces a subgoal `Coercible (Record r) (Record s)`?

Reply:
I think a unification error is fine, since it's likely an error anyway. For example you couldn't write:
```
test :: foral r s. { a :: Int | r } -> { a :: Int | s }
test = identity
```
The way to solve that would be a type equality that `r ~ s`.

Reply:
Right, that makes sense.

Reply:
Actually, maybe a unification error isn't right, and we should just punt the whole constraint? It's possible that it could be solved at the call site.

Reply:
I'm not sure I have fully appreciated the difference yet, to be honest.

Reply:
At the call-site `r` could unify with `s`, which would make the constraint solvable.

Reply:
I mean, if we are in this part of the code already, we are trying to solve a Coercible constraint, does that not mean we are already at the call site? I'm not sure I understand how/why we have a choice.

Reply:
Given
```purescript
test :: forall r s. { a :: Int | r } -> { a :: Id1 Int | s }
test = coerce
```
We are solving the instantiation of `coerce`. With the provided signature, the constraint is not solvable because `r` does not unify with `s`. We shouldn't throw a unification error because it's possible that it could be solved, given that `r` _does_ unify with `s`. That is, this is potentially solvable:
```purescript
test :: forall r s. Coercible { a :: Int | r } { a :: Id1 Int | s } => { a :: Int | r } -> { a :: Id1 Int | s }
test = coerce

test2 = test { a: 42 } :: { a :: Id1 Int }
```
because now `r` and `s` unify with `()`. We should not throw a unification error, we should just defer the constraint.

Reply:
Note that I think it's fine to throw a unification error for the existing case as written, because we know for a fact there is no solution, since there are concrete rows the differ. I'm only talking about polymorphic tails.

Reply:
Right, I see, thanks. That approach seems a bit odd to me because then `test` effectively has an invisible `Coercible` constraint, right? If I wrote a type annotation `forall r s. { a :: Int | r } -> { a :: Id1 Int | s }` then I would want the compiler to tell me if that function can’t actually be used for all possible choices of `r` and `s`; this feels like a parametricity violation to me.

Reply:
I'm not sure what you mean by "an invisible Coercible constraint". I'm just saying that the compiler should not fail with a unification error, it should fail saying that it can't solve the constraint, and so the user should add the constraint.

Reply:
Oh! I see now. That sounds reasonable to me, yes.